### PR TITLE
Add support for plan branches and overloading plan

### PR DESCRIPTION
**Pull request by emilingerslev**
With this PR shuttle will support setting the branch of a plan, like:
```
plan: git@github.com:lunarway/shuttle-example-go-plan.git#change-build
```
The `shuttle` CLI also supports 2 other new features to help develop
plans. Both is about overloading the current plan:

* A new flag `--plan`
* An environment variable `SHUTTLE_PLAN_OVERLOAD`

content of these two are handled the same, but the flag will take precedence
over the environment variable.

Setting the overload to a full plan specifier will change the plan fully. This
allows us to fx point a project to same plan in another directory, thus we can
develop one plan easily for multiple project.
Setting the overload to `#some-branch`, `#some-tag`, `#2b52c21` will change a
git plan's to target branch, tag or sha.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,8 +92,26 @@ $ shuttle run build tag=v1
...
+### Git Plan
+When using a git plan a url should look like:
+
+* `https://github.com/lunarway/shuttle-example-go-plan.git`
+* `git@github.com:lunarway/shuttle-example-go-plan.git`
+* `https://github.com/lunarway/shuttle-example-go-plan.git#change-build`
+* `git@github.com:lunarway/shuttle-example-go-plan.git#change-build`
+
+The `#change-build` points the plan to a specific branch, which by default would be `master`.
+It can also be used to point to a tag or a git SHA.
```
Comment (Crevil) [inline README.md]: Could you add an example of the Git SHA and tags as well?

Judging by the next section I expect it to be like `shuttle-example-go-plan.git#my-tag`? If this is the case, what happens if a tag and branch name conflicts? E.g. a `next` branch with a `next` tag?

File: cmd/root.go [hunk 2]
```diff
@@ -55,6 +56,11 @@ func init() {
 	rootCmd.PersistentFlags().StringVarP(&projectPath, "project", "p", ".", "Project path")
 	rootCmd.PersistentFlags().BoolVarP(&clean, "clean", "c", false, "Start from clean setup")
 	rootCmd.PersistentFlags().BoolVar(&skipGitPlanPulling, "skip-pull", false, "Skip git plan pulling step")
+	rootCmd.PersistentFlags().StringVar(&plan, "plan", "", `Overload the plan used.
+Specifying a local path with either a absolute path (/some/plan) or a relative path (../some/plan) to another location
```
Comment (Crevil) [inline cmd/root.go]:
```suggestion
Specifying a local path with either an absolute path (/some/plan) or a relative path (../some/plan) to another location
```

File: pkg/config/planargument.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"github.com/lunarway/shuttle/pkg/git"
+	"os"
+	"path"
+	"strings"
+)
+
+func isPlanArgumentAFilePlan(planArgument string) bool {
+	return strings.Index(planArgument, "/") == 0 || strings.Index(planArgument, "./") == 0 || strings.Index(planArgument, "../") == 0
```
Comment (Crevil) [inline pkg/config/planargument.go]: Could be implemented with [`strings.HasPrefix()`](https://golang.org/pkg/strings/#HasPrefix) for readability.

```suggestion
	return strings.HasPrefix(planArgument, "/") || strings.HasPrefix(planArgument, "./") || strings.HasPrefix(planArgument, "../")
```

File: pkg/config/planargument.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+func isPlanArgumentAPlan(planArgument string) bool {
+	return planArgument != "" && (git.IsGitPlan(planArgument) || isPlanArgumentAFilePlan(planArgument))
+}
+
+func getPlanFromPlanArgument(planArgument string) string {
+	switch {
+	case isPlanArgumentAFilePlan(planArgument) && isFilePath(planArgument, true):
```
Comment (Crevil) [inline pkg/config/planargument.go]: Could this not be omitted as it results in the same result as the `default` case, or am I missing something?

File: pkg/config/planargument.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+		return planArgument
+	case isPlanArgumentAFilePlan(planArgument) && isFilePath(planArgument, false):
+		wd, err := os.Getwd()
+		if err != nil {
+			panic(err)
```
Comment (Crevil) [inline pkg/config/planargument.go]: This ain't nice. Our error handling is in general not good and I would prefer this to return an error instead as it makes future improvements easier. My point is, new functions should make the error handling up to the caller instead.

Do you agree? If not just leave it and we can do an error handling overhaul in the near future.

File: pkg/config/shuttleconfig.go [hunk 6]
```diff
@@ -6,7 +6,7 @@ import (
 	"path"
 
 	"github.com/lunarway/shuttle/pkg/ui"
-	yaml "gopkg.in/yaml.v2"
+	"gopkg.in/yaml.v2"
```
Comment (Crevil) [inline pkg/config/shuttleconfig.go]: What is the purpose of this change?

File: pkg/config/shuttleplan.go [hunk 7]
```diff
@@ -67,23 +67,27 @@ func (p *ShuttlePlanConfiguration) Load(planPath string) *ShuttlePlanConfigurati
...
 	case plan == "":
 		uii.Verboseln("Using no plan")
 		return ""
 	case git.IsGitPlan(plan):
 		uii.Verboseln("Using git plan at '%s'", plan)
-		return git.GetGitPlan(plan, localShuttleDirectoryPath, uii, skipGitPlanPulling)
+		return git.GetGitPlan(plan, localShuttleDirectoryPath, uii, skipGitPlanPulling, planArgument)
 	case isMatching("^http://|^https://", plan):
-		panic("plan not valid: http is not supported yet")
+		panic(fmt.Sprintf("Plan '%v' is not valid: non-git http/https is not supported yet", plan))
```
Comment (Crevil) [inline pkg/config/shuttleplan.go]: If we explicitly haven't implemented HTTP(S) support, I think we should omit it in the documentation as well? This is confusing to me.

File: pkg/git/main.go [hunk 8]
```diff
@@ -64,8 +70,14 @@ func IsGitPlan(plan string) bool {
 }
 
 // GetGitPlan will pull git repository and return its path
-func GetGitPlan(plan string, localShuttleDirectoryPath string, uii ui.UI, skipGitPlanPulling bool) string {
+func GetGitPlan(plan string, localShuttleDirectoryPath string, uii ui.UI, skipGitPlanPulling bool, planArgument string) string {
 	parsedGitPlan := parseGitPlan(plan)
+
+	if strings.Index(planArgument, "#") == 0 {
```
Comment (Crevil) [inline pkg/git/main.go]: One could use `strings.HasPrefix()` again if one likes 👀 

```suggestion
	if strings.HasPrefix(planArgument, "#") {
```

File: pkg/git/main.go [hunk 9]
```diff
@@ -89,10 +101,17 @@ func GetGitPlan(plan string, localShuttleDirectoryPath string, uii ui.UI, skipGi
...
-			uii.Infoln("Pulling latest plan changes")
+			gitCmd("fetch origin", planPath, uii)
+			gitCmd(fmt.Sprintf("checkout %s", parsedGitPlan.head), planPath, uii)
+			status := getStatus(planPath)
+			if !status.isDetached {
+				uii.Infoln("Pulling latest plan changes on %v", parsedGitPlan.head)
+				gitCmd("pull origin", planPath, uii)
+				status = getStatus(planPath)
+			} else {
+				uii.EmphasizeInfoln("Skipping plan pull because its running on detached head")
```
Comment (Crevil) [inline pkg/git/main.go]: What does this mean? When can I get into this case? If I specify a specific commit SHA or?

File: tests.sh [hunk 10]
```diff
@@ -128,5 +128,34 @@ test_template_local_path() {
   assertErrorCode 0 -p examples/moon-base template ../custom-template.tmpl -o Dockerfile-custom
 }
 
+test_run_repo_say_branch() {
+  result=$(./shuttle -p examples/repo-project-branched run say)
+  if [[ ! "$result" =~ "something clever" ]]; then
+    fail "Expected output to contain 'something minor', but it was:\n$result"
```
Comment (Crevil) [inline tests.sh]: The `if` looks for `something clever` but this error message states `something minor`.

Looking into [shuttle-example-go-plan](https://github.com/lunarway/shuttle-example-go-plan/blob/9e5bec02d6bd737e6423d873d95fdab65e7814df/plan.yaml#L23) this should in fact be as you assert.

```suggestion
    fail "Expected output to contain 'something clever', but it was:\n$result"
```

File: tests.sh [hunk 11]
```diff
@@ -128,5 +128,34 @@ test_template_local_path() {
...
+  fi
+}
+
+test_run_repo_say() {
+  result=$(./shuttle -p examples/repo-project run say)
+  if [[ ! "$result" =~ "something masterly" ]]; then
+    fail "Expected output to contain 'something', but it was:\n$result"
```
Comment (Crevil) [inline tests.sh]: As above the error description does not match you expectations.

```suggestion
    fail "Expected output to contain 'something masterly', but it was:\n$result"
```

File: tests.sh [hunk 12]
```diff
@@ -128,5 +128,34 @@ test_template_local_path() {
...
+  fi
+}
+
+test_run_repo_say_tagged() {
+  result=$(./shuttle -p examples/repo-project --plan "#tagged" run say)
+  if [[ ! "$result" =~ "something tagged" ]]; then
+    fail "Expected output to contain 'something', but it was:\n$result"
```
Comment (Crevil) [inline tests.sh]: As above

```suggestion
    fail "Expected output to contain 'something tagged', but it was:\n$result"
```

File: README.md (hunk 1 above)
Comment (emilingerslev) [reply README.md]: Not totally sure what will happen, but it will defaults to git's behavior. I think that means that branches has precedence over the tags.

Comment (Crevil) [reply README.md]: All right. It might be an unlikely scenario, so let's just leave it as is :)

File: pkg/config/shuttleconfig.go (hunk 6 above)
Comment (emilingerslev) [reply pkg/config/shuttleconfig.go]: I think that dependency optimizer did that, due to the package is actually called "yaml" which makes this redundant.

File: pkg/config/shuttleplan.go (hunk 7 above)
Comment (emilingerslev) [reply pkg/config/shuttleplan.go]: We have it, just only for git repositories.

File: pkg/config/planargument.go (hunk 4 above)
Comment (emilingerslev) [reply pkg/config/planargument.go]: Maybe you are right, since `isFilePath(planArgument, false)` should not match this case.

File: pkg/config/planargument.go (hunk 5 above)
Comment (emilingerslev) [reply pkg/config/planargument.go]: I would agree with you, except in this case. This is extremely unlikely to fail, which is why I didn't want to expand the api with an error.

File: pkg/git/main.go (hunk 9 above)
Comment (emilingerslev) [reply pkg/git/main.go]: When you checkout a SHA or tag git will get into detached mode. This is what this means.
Not sure how to explain it further without getting into a more complex sentence.

Comment (Crevil) [reply pkg/git/main.go]: It's fine. The question was out of curiosity. :)
